File: skills/clawhub-git-sync-agent/scripts/sync_agent.py
#!/usr/bin/env python3
"""
Permanenter ClawHub ↔ Git Sync Agent
Multi-Node fähig, stündliche Ausführung
"""

import os
import sys
import json
import subprocess
from pathlib import Path
from datetime import datetime

# Import sync functions
sys.path.append('/home/openclaw/.openclaw/workspace/scripts')
from sync_clawhub_git import sync_to_git, sync_to_clawhub, log, validate_skill, get_file_hash, iter_sync_files

CLAWHUB_DIR = Path("/home/openclaw/.openclaw/workspace/skills")
GIT_DIR = Path("/home/openclaw/.openclaw/workspace/git/skills")
STATE_FILE = Path("/home/openclaw/.openclaw/workspace/db/sync_state.json")

def load_state():
    """Lädt den Sync-State"""
    if STATE_FILE.exists():
        with open(STATE_FILE, 'r') as f:
            return json.load(f)
    return {"sync_history": [], "pending": []}

def save_state(state):
    """Speichert den Sync-State"""
    # Ensure the parent directory exists (handle symlink to existing directory)
    if not STATE_FILE.parent.is_dir():
        STATE_FILE.parent.mkdir(parents=True, exist_ok=True)
    with open(STATE_FILE, 'w') as f:
        json.dump(state, f, indent=2)

def get_all_skills():
    """Findet alle Skills in beiden Verzeichnissen"""
    clawhub_skills = {
        d.name
        for d in CLAWHUB_DIR.iterdir()
        if d.is_dir() and not d.name.startswith('.') and (d / "SKILL.md").exists()
    }
    git_skills = {
        d.name
        for d in GIT_DIR.iterdir()
        if d.is_dir() and not d.name.startswith('.') and (d / "SKILL.md").exists()
    }
    return clawhub_skills.union(git_skills)

def init_git_repo(skill_path: Path, skill_name: str):
    """Initialisiert Git-Repo wenn nötig"""
    git_dir = skill_path / ".git"
    if not git_dir.exists():
        os.chdir(skill_path)
        subprocess.run(["git", "init"], capture_output=True)
        subprocess.run(["git", "add", "."], capture_output=True)
        subprocess.run(["git", "commit", "-m", f"Initial commit: {skill_name} skill"], capture_output=True)
        log(f"Git initialized for {skill_name}")

def sync_skill_bidirectional(skill_name: str):
    """Bidirektionale Synchronisation eines Skills"""
    clawhub_path = CLAWHUB_DIR / skill_name
    git_path = GIT_DIR / skill_name
    
    # Fall 1: Nur in ClawHub → zu Git
    if clawhub_path.exists() and not git_path.exists():
        log(f"NEW in ClawHub: {skill_name} → syncing to Git")
        if sync_to_git(skill_name, dry_run=False):
            init_git_repo(git_path, skill_name)
            return "synced_to_git"
    
    # Fall 2: Nur in Git → zu ClawHub
    elif git_path.exists() and not clawhub_path.exists():
        log(f"NEW in Git: {skill_name} → syncing to ClawHub")
        if sync_to_clawhub(skill_name, dry_run=False):
            return "synced_to_clawhub"
    
    # Fall 3: In beiden vorhanden
    elif clawhub_path.exists() and git_path.exists():
        if not validate_skill(clawhub_path):
            log(f"Validation failed for ClawHub skill: {skill_name}", "ERROR")
            return "error"
        if not validate_skill(git_path):
            log(f"Validation failed for Git skill: {skill_name}", "ERROR")
            return "error"

        clawhub_changes = preview_changes(clawhub_path, git_path)
        git_changes = preview_changes(git_path, clawhub_path)

        if not clawhub_changes and not git_changes:
            log(f"Content is identical for: {skill_name}")
            return "no_change"

        if clawhub_changes and not git_changes:
            log(f"Content difference detected for: {skill_name}")
            log(f"UPDATE: {skill_name} ClawHub content is newer or different → syncing to Git")
            if sync_to_git(skill_name, dry_run=False):
                os.chdir(git_path)
                subprocess.run(["git", "add", "."], capture_output=True)
                subprocess.run(["git", "commit", "-m", f"Sync from ClawHub content diff: {datetime.now().strftime('%Y-%m-%d %H:%M')}"], capture_output=True)
                return "updated_git"
            log(f"Failed to sync {skill_name} to Git after content diff", "ERROR")
            return "error"

        if git_changes and not clawhub_changes:
            log(f"Content difference detected for: {skill_name}")
            log(f"UPDATE: {skill_name} Git content is newer or different → syncing to ClawHub")
            if sync_to_clawhub(skill_name, dry_run=False):
                return "updated_clawhub"
            log(f"Failed to sync {skill_name} to ClawHub after content diff", "ERROR")
            return "error"

        log(f"Content difference detected for: {skill_name}")
        if newest_mtime(clawhub_path) >= newest_mtime(git_path):
            log(f"UPDATE: {skill_name} ClawHub content is newer or different → syncing to Git")
            if sync_to_git(skill_name, dry_run=False):
                os.chdir(git_path)
                subprocess.run(["git", "add", "."], capture_output=True)
                subprocess.run(["git", "commit", "-m", f"Sync from ClawHub content diff: {datetime.now().strftime('%Y-%m-%d %H:%M')}"], capture_output=True)
                return "updated_git"
        else:
            log(f"UPDATE: {skill_name} Git content is newer or different → syncing to ClawHub")
            if sync_to_clawhub(skill_name, dry_run=False):
                return "updated_clawhub"

        log(f"Failed to resolve content diff for {skill_name}", "ERROR")
        return "error"
    
    return "no_change"

# --- Hinzufügen dieser Hilfsfunktion ---
def get_hashes(skill_dir: Path):
    """Erzeugt ein Dictionary von Datei-Hashes für einen Skill-Ordner."""
    hashes = {}
    for file_path, rel_path in iter_sync_files(skill_dir):
        hashes[str(rel_path)] = get_file_hash(file_path)
    return hashes

def preview_changes(source_dir: Path, target_dir: Path):
    """Berechnet Sync-Änderungen in einer Richtung, ohne zu schreiben."""
    changes = []
    for src_file, rel_path in iter_sync_files(source_dir):
        tgt_file = target_dir / rel_path
        if not tgt_file.exists():
            changes.append(f"ADD {rel_path}")
        elif get_file_hash(src_file) != get_file_hash(tgt_file):
            changes.append(f"UPDATE {rel_path}")
    return changes

def newest_mtime(skill_dir: Path) -> float:
    """Ermittelt die neueste mtime über alle relevanten Dateien."""
    mtimes = [file_path.stat().st_mtime for file_path, _ in iter_sync_files(skill_dir)]
    return max(mtimes) if mtimes else 0.0

def main():
    """Hauptfunktion des Sync-Agents mit Dry-Run Phase"""
    log("=== ClawHub ↔ Git Sync Agent gestartet ===")
    
    # Load previous state
    state = load_state()
    all_skills = get_all_skills()
    log(f"Gefundene Skills: {len(all_skills)}")
    
    # Dry-Run Phase: only report changes, no actual modifications
    log("--- Dry-Run Phase Start ---")
    for skill in sorted(all_skills):
        # Perform dry-run sync in both directions to capture potential changes
        sync_to_git(skill, dry_run=True)
        sync_to_clawhub(skill, dry_run=True)
    log("--- Dry-Run Phase End ---")
    
    results = {
        "synced_to_git": [],
        "synced_to_clawhub": [],
        "updated_git": [],
        "updated_clawhub": [],
        "no_change": [],
        "errors": []
    }
    
    # Actual Sync Phase
    for skill in sorted(all_skills):
        try:
            result = sync_skill_bidirectional(skill)
            results[result].append(skill)
        except Exception as e:
            log(f"ERROR syncing {skill}: {e}", "ERROR")
            results["errors"].append(skill)
    
    # Zusammenfassung
    log("\n=== SYNC ZUSAMMENFASSUNG ===")
    log(f"Neu in Git: {len(results['synced_to_git'])} - {results['synced_to_git']}")
    log(f"Neu in ClawHub: {len(results['synced_to_clawhub'])} - {results['synced_to_clawhub']}")
    log(f"Git aktualisiert: {len(results['updated_git'])} - {results['updated_git']}")
    log(f"ClawHub aktualisiert: {len(results['updated_clawhub'])} - {results['updated_clawhub']}")
    log(f"Keine Änderung: {len(results['no_change'])}")
    log(f"Fehler: {len(results['errors'])} - {results['errors']}")
    
    # State speichern
    if "sync_history" not in state:
        state["sync_history"] = []
    state["sync_history"].append({
        "timestamp": datetime.now().isoformat(),
        "results": results
    })
    # Nur letzte 100 Einträge behalten
    state["sync_history"] = state["sync_history"][-100:]
    save_state(state)
    
    log("=== Sync Agent beendet ===\n")

if __name__ == "__main__":
    main()

File: javascript/sync_agent.js
#!/usr/bin/env node
// sync_agent.py — portiert nach javascript
// Quelle: python, OpenClaw@gateway1:skills/clawhub-git-sync-agent/scripts/sync_agent.py
// Erzeugt: 2026-08-21 durch ABSTRACTIONS_MANAGER.py

/**
 * Permanenter ClawHub ↔ Git Sync Agent
 * Multi-Node fähig, stündliche Ausführung
 */

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');
const { createReadStream, createWriteStream } = require('fs');
const { createGzip } = require('zlib');
const { pipeline } = require('stream/promises');
const tar = require('tar-stream');

// Import sync functions
const syncModulePath = '/home/openclaw/.openclaw/workspace/scripts/sync_clawhub_git.js';
const { syncToGit, syncToClawhub, log, validateSkill, getFileHash } = require(syncModulePath);

const CLAWHUB_DIR = "/home/openclaw/.openclaw/workspace/skills";
const GIT_DIR = "/home/openclaw/.openclaw/workspace/git/skills";
const STATE_FILE = "/home/openclaw/.openclaw/workspace/db/sync_state.json";

// Root directory for backups
const BACKUP_ROOT = "/home/openclaw/.openclaw/workspace/backups/sync_agent";

function loadState() {
    /** Lädt den Sync-State */
    if (fs.existsSync(STATE_FILE)) {
        const data = fs.readFileSync(STATE_FILE, 'utf8');
        return JSON.parse(data);
    }
    return { sync_history: [], pending: [] };
}

function saveState(state) {
    /** Speichert den Sync-State */
    const dir = path.dirname(STATE_FILE);
    if (!fs.existsSync(dir)) {
        fs.mkdirSync(dir, { recursive: true });
    }
    fs.writeFileSync(STATE_FILE, JSON.stringify(state, null, 2));
}

function getAllSkills() {
    /** Findet nur valide Skill-Verzeichnisse in beiden Verzeichnissen. */
    const clawhubSkills = new Set();
    if (fs.existsSync(CLAWHUB_DIR)) {
        for (const item of fs.readdirSync(CLAWHUB_DIR)) {
            const fullPath = path.join(CLAWHUB_DIR, item);
            const stat = fs.statSync(fullPath);
            if (stat.isDirectory() && !item.startsWith('.') && !item.startsWith('_')) {
                if (fs.existsSync(path.join(fullPath, "SKILL.md"))) {
                    clawhubSkills.add(item);
                }
            }
        }
    }

    const gitSkills = new Set();
    if (fs.existsSync(GIT_DIR)) {
        for (const item of fs.readdirSync(GIT_DIR)) {
            const fullPath = path.join(GIT_DIR, item);
            const stat = fs.statSync(fullPath);
            if (stat.isDirectory() && !item.startsWith('.') && !item.startsWith('_')) {
                if (fs.existsSync(path.join(fullPath, "SKILL.md"))) {
                    gitSkills.add(item);
                }
            }
        }
    }

    return new Set([...clawhubSkills, ...gitSkills]);
}

function initGitRepo(skillPath, skillName) {
    /** Initialisiert Git-Repo wenn nötig */
    const gitDir = path.join(skillPath, ".git");
    if (!fs.existsSync(gitDir)) {
        process.chdir(skillPath);
        execSync("git init", { stdio: 'ignore' });
        execSync("git add .", { stdio: 'ignore' });
        execSync(`git commit -m "Initial commit: ${skillName} skill"`, { stdio: 'ignore' });
        log(`Git initialized for ${skillName}`);
    }
}

async function backupSkillDir(skillPath, skillName) {
    /** Creates a timestamped tar.gz backup of a skill directory. */
    if (!fs.existsSync(skillPath)) {
        return;
    }
    const timestamp = new Date().toISOString().replace(/[:.]/g, '').replace('T', '').substring(0, 12);
    const backupDir = path.join(BACKUP_ROOT, timestamp);
    if (!fs.existsSync(backupDir)) {
        fs.mkdirSync(backupDir, { recursive: true });
    }
    const archiveName = `${skillName}_${timestamp}.tar.gz`;
    const archivePath = path.join(backupDir, archiveName);
    
    // Create tar.gz manually since we don't have shutil equivalent
    const pack = tar.pack();
    const gzip = createGzip();
    const output = createWriteStream(archivePath);
    
    // Walk through directory and add files to tar
    function walkDirectory(currentPath, basePath) {
        const items = fs.readdirSync(currentPath);
        for (const item of items) {
            const fullPath = path.join(currentPath, item);
            const relativePath = path.relative(basePath, fullPath);
            const stat = fs.statSync(fullPath);
            
            if (stat.isDirectory()) {
                // Add directory entry
                pack.entry({ name: relativePath + '/' }, null, () => {});
                walkDirectory(fullPath, basePath);
            } else {
                // Add file entry
                const entry = pack.entry({ name: relativePath }, null);
                const fileStream = createReadStream(fullPath);
                fileStream.pipe(entry);
            }
        }
    }
    
    walkDirectory(skillPath, skillPath);
    pack.finalize();
    
    await pipeline(pack, gzip, output);
    log(`Backup created for ${skillName} at ${archivePath}`);
}

async function syncSkillBidirectional(skillName, dryRun = false) {
    /** Bidirektionale Synchronisation eines Skills */
    const clawhubPath = path.join(CLAWHUB_DIR, skillName);
    const gitPath = path.join(GIT_DIR, skillName);

    // Fall 1: Nur in ClawHub → zu Git
    if (fs.existsSync(clawhubPath) && !fs.existsSync(gitPath)) {
        log(`NEW in ClawHub: ${skillName} → syncing to Git`);
        if (!dryRun) {
            await backupSkillDir(clawhubPath, `${skillName}_clawhub`);
        }
        if (syncToGit(skillName, dryRun)) {
            if (!dryRun) {
                initGitRepo(gitPath, skillName);
            }
            return "synced_to_git";
        }
    }

    // Fall 2: Nur in Git → zu ClawHub
    else if (fs.existsSync(gitPath) && !fs.existsSync(clawhubPath)) {
        log(`NEW in Git: ${skillName} → syncing to ClawHub`);
        if (!dryRun) {
            await backupSkillDir(gitPath, `${skillName}_git`);
        }
        if (syncToClawhub(skillName, dryRun)) {
            return "synced_to_clawhub";
        }
    }

    // Fall 3: In beiden vorhanden → Vergleiche Timestamps
    else if (fs.existsSync(clawhubPath) && fs.existsSync(gitPath)) {
        // --- MODIFIZIERTE LOGIK: Robusterer Datei-Hash-Vergleich ---

        // Stelle sicher, dass beide als gültige Skills validiert werden
        if (!validateSkill(clawhubPath)) {
            log(`Validation failed for ClawHub skill: ${skillName}`, "ERROR");
            return "error";
        }
        if (!validateSkill(gitPath)) {
            log(`Validation failed for Git skill: ${skillName}`, "ERROR");
            return "error";
        }

        // Berechne Hashes für clawhub und git
        const clawhubHashes = getHashes(clawhubPath);
        const gitHashes = getHashes(gitPath);

        if (JSON.stringify(clawhubHashes) !== JSON.stringify(gitHashes)) {
            log(`Content difference detected for: ${skillName}`);

            // Einfache (aber oft ausreichende) Logik: Wenn clawhub neuer ist, lade hoch.
            // Eine detailliertere Strategie (z.B. welche Version von Git übernehmen)
            // könnte hier implementiert werden, falls nötig.
            // Für jetzt: Wenn sie sich unterscheiden, priorisieren wir ClawHub > Git
            // und aktualisieren Git.

            const clawhubStat = fs.statSync(clawhubPath);
            const gitStat = fs.statSync(gitPath);
            const direction = clawhubStat.mtimeMs >= gitStat.mtimeMs ? "to-git" : "to-clawhub";
            log(`UPDATE: ${skillName} → syncing ${direction}`);
            if (!dryRun) {
                await backupSkillDir(clawhubPath, `${skillName}_clawhub`);
                await backupSkillDir(gitPath, `${skillName}_git`);
            }
            const sync = direction === "to-git" ? syncToGit : syncToClawhub;
            if (sync(skillName, dryRun)) {
                if (!dryRun && direction === "to-git") {
                    process.chdir(gitPath);
                    execSync("git add .", { stdio: 'ignore' });
                    execSync(`git commit -m "Sync from ClawHub content diff: ${new Date().toISOString().slice(0, 16).replace('T', ' ')}"`, { stdio: 'ignore' });
                }
                return direction === "to-git" ? "updated_git" : "updated_clawhub";
            } else {
                log(`Failed to sync ${skillName} to Git after content diff`, "ERROR");
                return "error";
            }
        } else {
            log(`Content is identical for: ${skillName}`);
            return "no_change";
        }
    }

    return "no_change";
}

// --- Hinzufügen dieser Hilfsfunktion ---
function getHashes(skillDir) {
    /** Erzeugt ein Dictionary von Datei-Hashes für einen Skill-Ordner. */
    const hashes = {};
    
    function walk(currentPath) {
        const items = fs.readdirSync(currentPath);
        for (const item of items) {
            const fullPath = path.join(currentPath, item);
            const stat = fs.statSync(fullPath);
            
            if (stat.isDirectory()) {
                // Skip .git directories
                if (item !== '.git') {
                    walk(fullPath);
                }
            } else {
                // Skip .git files
                if (!fullPath.includes('.git')) {
                    const relativePath = path.relative(skillDir, fullPath);
                    hashes[relativePath] = getFileHash(fullPath);
                }
            }
        }
    }
    
    walk(skillDir);
    return hashes;
}

async function main() {
    /** Hauptfunktion des Sync-Agents */
    const args = process.argv.slice(2);
    const DRY_RUN = args.includes('--dry-run');
    log("=== ClawHub ↔ Git Sync Agent gestartet ===");

    const state = loadState();
    const allSkills = getAllSkills();
    log(`Gefundene Skills: ${allSkills.size}`);

    const results = {
        synced_to_git: [],
        synced_to_clawhub: [],
        updated_git: [],
        updated_clawhub: [],
        no_change: [],
        errors: []
    };

    for (const skill of Array.from(allSkills).sort()) {
        try {
            const result = await syncSkillBidirectional(skill, DRY_RUN);
            results[result].push(skill);
        } catch (e) {
            log(`ERROR syncing ${skill}: ${e.message}`, "ERROR");
            results.errors.push(skill);
        }
    }

    // Zusammenfassung
    log("\n=== SYNC ZUSAMMENFASSUNG ===");
    log(`Neu in Git: ${results.synced_to_git.length} - ${results.synced_to_git}`);
    log(`Neu in ClawHub: ${results.synced_to_clawhub.length} - ${results.synced_to_clawhub}`);
    log(`Git aktualisiert: ${results.updated_git.length} - ${results.updated_git}`);
    log(`ClawHub aktualisiert: ${results.updated_clawhub.length} - ${results.updated_clawhub}`);
    log(`Keine Änderung: ${results.no_change.length}`);
    log(`Fehler: ${results.errors.length} - ${results.errors}`);

    // Ein Dry-Run bleibt vollständig nicht-mutierend (abgesehen vom Audit-Log).
    if (!DRY_RUN) {
        if (!state.sync_history) {
            state.sync_history = [];
        }
        state.sync_history.push({
            timestamp: new Date().toISOString(),
            results: results
        });
        // Nur letzte 100 Einträge behalten
        state.sync_history = state.sync_history.slice(-100);
        saveState(state);
    }

    log("=== Sync Agent beendet ===\n");
}

if (require.main === module) {
    main().catch(err => {
        console.error(err);
        process.exit(1);
    });
}
